### PR TITLE
Use VMs

### PR DESCRIPTION
**Pull request by jim-minter**

### CODE REVIEW

**Comment by jim-minter** [comment]
@pweil- @kargakis @kwoodson @mjudeikis ptal

File: pkg/arm/data/azuredeploy.json [hunk 1]
```diff
@@ -171,18 +254,21 @@
                                             "properties": {
                                                 "primary": true,
                                                 "subnet": {
-{{ if $app.VnetSubnetID }}
-                                                    "id": "{{ $app.VnetSubnetID }}"
-{{ else }}
-                                                    "id": "[concat(resourceId('Microsoft.Network/virtualNetworks', 'vnet'), '/subnets/default')]"
-{{ end }}
+                                                    "id": "{{if $app.VnetSubnetID}}{{$app.VnetSubnetID}}{{else}}[concat(resourceId('Microsoft.Network/virtualNetworks', 'vnet'), '/subnets/default')]{{end}}"
```
Comment (mjudeikis) [inline pkg/arm/data/azuredeploy.json]: +1

File: pkg/config/generate.go [hunk 2]
```diff
@@ -121,9 +119,12 @@ func selectContainerImages(cs *acsapi.ContainerService, c *Config) {
 	}
 }
 
+func internalAPIServerHostname(cs *acsapi.ContainerService) string {
```
Comment (mjudeikis) [inline pkg/config/generate.go]: I assume this will change when we get either dedicated zone or some identifier.

File: pkg/addons/client.go [hunk 3]
```diff
@@ -109,11 +116,22 @@ func (c *client) waitForHealthz() error {
 
 	for {
 		resp, err := cli.Do(req)
-		if err, ok := err.(*url.Error); ok && (err.Timeout() || err.Err == io.EOF || err.Err == io.ErrUnexpectedEOF) {
-			time.Sleep(time.Second)
-			continue
+		if err, ok := err.(*url.Error); ok {
```
Comment (kargakis) [inline pkg/addons/client.go]: Create a helper to keep this loop cleaner
```go
func needsRetry(err error) bool {
	if err, ok := err.(*url.Error); ok {
		if err, ok := err.Err.(*net.OpError); ok {
			if err, ok := err.Err.(*os.SyscallError); ok {
				if err.Err == syscall.ENETUNREACH {
					return true
				}
			}
		}
		if err.Timeout() || err.Err == io.EOF || err.Err == io.ErrUnexpectedEOF {
			return true
		}
		return false
	}
}
```

File: pkg/arm/data/azuredeploy.json [hunk 4]
```diff
@@ -26,7 +26,93 @@
                 ]
             }
         },
-{{ end }}
+{{- end }}
+{{/* TODO: we may be able to avoid the IP and LB here if we stop using */}}
+{{/* DaemonSets for SDN.  Then Master nodes should come up as Ready */}}
```
Comment (kargakis) [inline pkg/arm/data/azuredeploy.json]: So how would you deploy SDN in this case?

File: pkg/config/generate.go [hunk 5]
```diff
@@ -409,7 +410,7 @@ func Generate(cs *acsapi.ContainerService, c *Config) (err error) {
 			// sync kubeconfig has the same capabilities as admin kubeconfig, only difference
 			// is the use of HCP internal DNS to avoid waiting for the Azure loadbalancer to
```
Comment (kargakis) [inline pkg/config/generate.go]: Follow-up issue to remove this in favor of using a service account.

File: glide.yaml [hunk 6]
```diff
@@ -3,33 +3,3 @@ import:
...
-  - ssh
-- package: k8s.io/apimachinery
-  subpackages:
-  - pkg/api/errors
-  - pkg/api/meta
-  - pkg/apis/meta/v1
-  - pkg/apis/meta/v1/unstructured
-  - pkg/runtime/schema
-  - pkg/util/wait
-- package: k8s.io/client-go
```
Comment (kargakis) [inline glide.yaml]: Why are these removed?

File: pkg/arm/data/azuredeploy.json (hunk 4 above)
Comment (jim-minter) [reply pkg/arm/data/azuredeploy.json]: static pods

File: pkg/config/generate.go (hunk 5 above)
Comment (jim-minter) [reply pkg/config/generate.go]: not sure I understand:
a) sync effectively needs system:admin, so it might as well just use it
b) there's a bootstrapping issue whereby the first time sync runs, all there is is system:admin.

File: pkg/arm/data/master-startup.sh [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+rm -rf /etc/etcd/* /etc/origin/master/*
+
+base64 -d <<< {{ CertAsBytes .Config.EtcdCaCert | Base64Encode }} >/etc/etcd/ca.crt
+base64 -d <<< {{ CertAsBytes .Config.EtcdServerCert | Base64Encode }} >/etc/etcd/server.crt
+base64 -d <<< {{ PrivateKeyAsBytes .Config.EtcdServerKey | Base64Encode }} >/etc/etcd/server.key
+base64 -d <<< {{ CertAsBytes .Config.EtcdPeerCert | Base64Encode }} >/etc/etcd/peer.crt
+base64 -d <<< {{ PrivateKeyAsBytes .Config.EtcdPeerKey | Base64Encode }} >/etc/etcd/peer.key
+
+base64 -d <<< {{ YamlMarshal .Config.AdminKubeconfig | Base64Encode }} >/etc/origin/node/node.kubeconfig
+base64 -d <<< {{ CertAsBytes .Config.CaCert | Base64Encode }} >/etc/origin/node/client-ca.crt
```
[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/arm/data/master-startup.sh]: thanks!

File: pkg/config/generate.go (hunk 5 above)
Comment (kargakis) [reply pkg/config/generate.go]: Yeah, you are right about the boostrapping problem.

File: glide.yaml (hunk 6 above)
Comment (jim-minter) [reply glide.yaml]: They're all transitively implied by the acs-engine import, which is currently pinning a lot of versions.  That's a pretty horrible situation and this removal helps to make that evident.  We'll want to somehow cut that out so we can decide the versions of the libraries we need, not have it decided for us on the basis of importing a single API package :-/

Comment (kargakis) [reply glide.yaml]: That's pretty bad, we use client-go directly in `sync` so it's not a transitive dependency and certain client-go versions map to certain k8s versions.

File: pkg/addons/client.go (hunk 3 above)
Comment (jim-minter) [reply pkg/addons/client.go]: yes, it was about time to do that ;-)

File: glide.yaml (hunk 6 above)
Comment (mjudeikis) [reply glide.yaml]: What about if we kick-off separate vendor tree/sub project for projects not dependent on acs-engine (like sync)?

**Comment by jim-minter** [comment]
Thanks for the reviews.  Addressed.  Merging.

File: pkg/addons/data/StatefulSet.apps/openshift-metrics/prometheus.yaml [hunk 8]
```diff
@@ -37,7 +37,6 @@ spec:
         - -openshift-ca=/var/run/secrets/kubernetes.io/serviceaccount/ca.crt
         - -skip-auth-regex=^/metrics
         image: '*** GENERATED ***'
-        imagePullPolicy: Always
```
Comment (kwoodson) [inline pkg/addons/data/StatefulSet.apps/openshift-metrics/prometheus.yaml]: for my understanding, why did these go away?  Looks like the default is to `IfNotPresent` which causes the kubelet to skip pulling if it exists vs forcing a pull each time.

Src:
https://kubernetes.io/docs/concepts/containers/images/

File: create.sh [hunk 9]
```diff
@@ -57,13 +52,18 @@ properties:
   publicHostname: openshift.$RESOURCEGROUP.$DNS_DOMAIN
   routingConfigSubdomain: $RESOURCEGROUP.$DNS_DOMAIN
   agentPoolProfiles:
-  - name: compute
-    count: 1
+  - name: master
+    role: master
+    count: 3
```
[the remark this answers is not in the document]
Comment (jim-minter) [reply create.sh]: in due course this should be more flexible https://github.com/openshift/openshift-azure/issues/119

File: pkg/addons/translate.go [hunk 10]
```diff
@@ -343,19 +356,30 @@ var Translations = map[string][]struct {
 	},
 	"Secret/kube-service-catalog/apiserver-ssl": {
 		{
-			Path:     jsonpath.MustCompile("$.stringData.'tls.crt'"),
+			Path:     jsonpath.MustCompile("$.stringData.'apiserver.crt'"),
```
[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/addons/translate.go]: no - the single quotes are required because of the period.  we translate the Secret "data" to "stringData" at https://github.com/openshift/openshift-azure/blob/master/pkg/addons/clean.go#L45 to save extra hassle with base64 encodings.

File: pkg/arm/data/master-startup.sh [hunk 11]
```diff
@@ -0,0 +1,570 @@
...
+resourceGroup: {{ .ContainerService.Properties.AzProfile.ResourceGroup | quote }}
+location: {{ .ContainerService.Location | quote }}
+securityGroupName: nsg-compute
+primaryScaleSetName: ss-compute
+vmType: vmss
+EOF
+
+# TODO: investigate the --manifest-url Kubelet parameter and see if it might
+# help us at all
+cat >/etc/origin/node/pods/etcd.yaml <<EOF
```
[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/arm/data/master-startup.sh]: The direction I'd like to take is to get these under our control in openshift-azure rather than openshift-ansible - that's why

File: pkg/addons/data/StatefulSet.apps/openshift-metrics/prometheus.yaml (hunk 8 above)
Comment (jim-minter) [reply pkg/addons/data/StatefulSet.apps/openshift-metrics/prometheus.yaml]: Yes, this isn't right. https://github.com/openshift/openshift-azure/issues/120.
